### DISCUSSION TITLE
Обсуждения

### DISCUSSION THREAD

**Discussion by nick-nh** [nick-nh/qlua#6] [General]
Здесь можно создавать дискуссии по функциональности, вопросам реализации, пожелания разработки и т.д.

**Comment by shtur2005** [comment]
Добрый день! А как пользоваться вашим роботом?
Вроде все настроил, но он не работает.
--Вот эти значения куда вводить?
Класс инструмента 'SPBFUT'
Код инструмента 'SRH2'
--Идентификатор счета и Код клиента это одно и то же значение?
'7683r51'
создал график с ТФ3 добавил к нему идентификатор 'Sheet11'
отредактировал время работы
при запуске ругается на строку 2262: attempt to index a nil value (global 'DS')
пишет что нет доступа к свечам sih1

**Reply by nick-nh** [reply]
Добрый день.

В описании написано, что алгоритмы - это подключаемые файлы. В них и есть настройки для каждого файла.
В самом файле скрипта задаются только умолчания.

**Comment by shtur2005** [comment]
спасибо. попробовал, не вышло пойду дальше. ничего страшного.

**Comment by CAHDPO** [comment]
Здравствуйте, хотел уточнить возможность реализации индикатора, который в подвале будет выводить в виде двух линий реальное ГО на покупку и продажу, которое будет рассчитываться по формулам из приложенного файла. https://disk.yandex.ru/i/La3cxW9qV8DyVg Так же есть часть кода с формулами https://disk.yandex.ru/d/KpBBwQAoh759OA

**Reply by nick-nh** [reply]
Добрый день. Не сложно, тем более что эта функция написана мной. Но не очень понятно как выводить данные, на каждый тик, т.к. расчет имеет смысл для конкретной цены?

**Reply by CAHDPO** [reply]
Здравствуйте, спасибо, что откликнулись, думаю, что выводить на каждый тик, а значение линии на графике фиксировать по закрытию свечи текущего таймфрейма.

**Comment by CAHDPO** [comment]
Приветствую, я тут понаблюдал, в принципе, особой нужды выводит данные каждый тик, я думаю нет.  Достаточно было бы обновления раз в 1сек., информативность линий мне кажется сильно не пострадала бы.

**Comment by 123DNK123** [comment]
вопрос не правильно задал как в qlua записать цену закрытия предыдущей свечи и значение ЕМА на предыдущей свече

**Comment by 123DNK123** [comment]
в tradingview  это записывается как close[1] и ema[1] а как в qlua

**Comment by nick-nh** [comment]
Наверно, все же, не записать, а прочитать.

В отличии от tradingview, который напрямую предоставляет контекст к данным, в терминале Квик данные необходимо заказывать с сервера, получать и обрабатывать. Не зря он называется терминал. Это комплексный вопрос.

Советую начать изучение методов:
документация от разработчика терминала
https://arqatech.com/upload/iblock/194/quik_lua.zip
https://arqatech.com/upload/iblock/398/INDICATORS.zip

Ну и в этом проекте почти все алгоритмы так или иначе получают данные, рассчитывают. Например простой алгоритм EMA
https://github.com/nick-nh/qlua/blob/master/robot/shiftMaAlgo.lua

**Comment by 123DNK123** [comment]
Большое спасибо за помощь))

**Comment by 123DNK123** [comment]
Добрый день. При запуске выдает такую ошибку  robotAlgo.lua:1364: attempt to index a nil value (global 'stv')
и тоже самое на robotAlgo.lua:2276: attempt to index a nil value (global 'stv'). подскажите пожалуйста с чем может быть связано.

**Comment by nick-nh** [comment]
Добрый день.
Это библиотека
require("StaticVar")

В описании есть ссылка на нее. Или просто найдите ее поиском. Она быстро найдется.

**Comment by 123DNK123** [comment]
error loading module 'StaticVar' from file 'C:\BCS_Work\QUIK_BCS\StaticVar.dll':
	%1 не является приложением Win32. библиотека скачаная по ссылке выдает такую ошибку. а скачаная на стороннем сайте 
C:\Program Files\Notepad++\robotAlgo.lua:1364: attempt to index a nil value (global 'stv')

**Reply by nick-nh** [reply]
Надо взять нужную версии сборки. Они разные для lua 5.1 5.3 и разные для win32 и x64

**Comment by 123DNK123** [comment]
stv = require("StaticVar") решилось таким образом)

**Comment by 123DNK123** [comment]
robotAlgo.lua:2116: attempt to index a nil value (local 'calcAlgoValue') теперь такую ошибку выдает . постоянно на что то натыкаюсь)

**Reply by nick-nh** [reply]
Потому что необходимо заполнить все данные в стратегии. Скрипт же выложен давно, там заполнены данные по инструментам, которые уже не торгуются.
Прочитайте внимательно описание к скрипту. Этот скрипт для обучения.

**Comment by 123DNK123** [comment]
Произошла ошибка при загрузке файла 'C:\Program Files\DNK\Pictures\sell.bmp'. в папке DNK лежат все файлы робота

**Comment by 123DNK123** [comment]
Добрый вечер. подключил свой модуль к роботу при СТАРТЕ выдает такую ошибку 
.lua:672: bad argument #1 to 'for iterator' (table expected, got nil) .подскажите с чем это может быть связано. Спасибо

**Comment by 123DNK123** [comment]
672  for k,v in pairs(presets[curPreset].settingsAlgo) do
  673     Settings[k] = v
  674      myLog('Установка Settings.'..k..' = '..tostring(v))
  675   end

**Reply by nick-nh** [reply]
Если в этом месте ошибка, то значит некорректно добавлены настройки в стратегии.
Т.е. нет таблицы presets[curPreset]

Обращаю внимание, что переменная curPreset по умолчанию задается в строке 192
curPreset = 4

**Comment by 123DNK123** [comment]
С таблицей все получилось Спасибо)
не могу понять что не так с этим вычислением
local upper2 = (DS:H(index)+DS:H(index-1)+ DS:H(index-2)+DS:H(index-3)+DS:H(index-4))/5
local basis22 = (upper2+upper2(index-1)+upper2(index-2)+upper2(index-3)+upper2(index-4))

вторая строка выдает такую ошибку :attempt to call a number value (local 'upper2')

**Reply by nick-nh** [reply]
Ну так нотация () - это вызов функции: upper2(index-1)
но upper2 - это явно число

**Comment by 123DNK123** [comment]
что то не соображу как это записать чтобы  upper2  определялся как число а не вызов функции. можете подсказать или где можно почитать. Спасибо

**Reply by nick-nh** [reply]
Для этого необходимо пронять что значит upper2(index-1)

**Comment by 123DNK123** [comment]
как пример ЕМА -это на текущем баре, а ЕМА(index-1)- это с теми же параметрами но на предыдущем баре .

**Reply by nick-nh** [reply]
ЕМА - это функция. А если речь про таблицу (массив), то ее необходимо создать, зафиксировать в ней значение, и уже потом извлекать, но используя квадратные скобки []

где-то объявить upper2 

local upper2  = {}

......

А уже в месте расчета присвоит по индексу значение

upper2[index] = (DS:H(index)+DS:H(index-1)+ DS:H(index-2)+DS:H(index-3)+DS:H(index-4))/5

а потом уже использовать:

local basis22 = (upper2[index]+upper2[index-1]+upper2[index-2]+upper2[index-3]+upper2[index-4])

**Comment by 123DNK123** [comment]
upper2 = DS:H(index)+DS:H(index-1) это на текущем баре рассчитано
а upper2(index-1) это тот же upper2  только рассчитанный на предыдущем баре 
так правильнее объяснить) Спасибо

**Reply by nick-nh** [reply]
Это не tradingview. Все необходимо инициализировать, как во всех языках программирования.
Советую, все же, начать с изучения самого языка lua. qlua - это просто надстройка, не более. Сам язык lua такой, какой есть. Со всеми правилами.

**Comment by 123DNK123** [comment]
Понял большое Спасибо

**Comment by 123DNK123** [comment]
Добрый день) подскажите пожалуйста с чем может быть связана ошибка(выскакивает после открытия позиции) :
attempt to perform arithmetic on a string value (local 'AtPrice')

else -- открыт SELL

        if TAKE_PROFIT~=0 then
            if TakeProfitPrice == 0 then
                tp_stopprice	= round(AtPrice - TAKE_PROFIT*priceKoeff, SCALE) -- Уровень цены, когда активируется Тейк-профит

**Reply by nick-nh** [reply]
Добрый день. Если это функция getSLTP_Price, то проверьте кто ее вызывает и передает в параметрах вызова. Этих вызовов не так много.

**Comment by 123DNK123** [comment]
Добрый день.  При выполнении скрипта 
if Type == 'BUY' then
        if TAKE_PROFIT~=0 then
            if TakeProfitPrice == 0 then
                tp_stopprice	= round(AtPrice + TAKE_PROFIT*priceKoeff, SCALE) -- Уровень цены, когда активируется Тейк-профит
            elseif isPriceMove and shiftProfit then
                --tp_stopprice = round(TakeProfitPrice + (TRAILING_ACTIVATED and priceMoveVal or STOP_LOSS)*priceKoeff/2, SCALE)    -- немного сдвигаем тейк-профит
                tp_stopprice = round(TakeProfitPrice + priceMoveVal/2, SCALE)    -- немного сдвигаем тейк-профит
            else tp_stopprice = TakeProfitPrice
            end
        end
тейк профит выставляется в правильном направлении , но при этом выставляется как будто в скрипте указано 
  tp_stopprice	= round(AtPrice - TAKE_PROFIT*priceKoeff, SCALE) -- Уровень цены, когда активируется Тейк-профит
            elseif isPriceMove and shiftProfit then
                --tp_stopprice = round(TakeProfitPrice - (TRAILING_ACTIVATED and priceMoveVal or STOP_LOSS)*priceKoeff/2, SCALE)   

тоже самое и стоп лосс в обратную сторону . При этом расчет и выставление при  else -- открыт SELL происходит правильно.  в чем может быть загвоздка. подскажите пожалуйста.

**Reply by nick-nh** [reply]
Добрый день. 
Я не очень понял вопрос.

**Comment by 123DNK123** [comment]
Если совершилась покупка выставляется стоп лосс и тейк профит: в скрипте формулы
tp_stopprice = round(lastDealPrice + TAKE_PROFIT*priceKoeff, SCALE)
sl_stopprice = round(lastDealPrice- STOP_LOSS*priceKoeff, SCALE)
Но  выставляется как если бы формулы были такими 
tp_stopprice = round(lastDealPrice - TAKE_PROFIT*priceKoeff, SCALE)
sl_stopprice = round(lastDealPrice+ STOP_LOSS*priceKoeff, SCALE)
При этом направление выставления заявок на стоп и тейк правильное на продажу.
 Если совершилась продажа выставляется стоп лосс и тейк профит: в скрипте формулы
tp_stopprice = round(lastDealPrice - TAKE_PROFIT*priceKoeff, SCALE)
sl_stopprice = round(lastDealPrice +STOP_LOSS*priceKoeff, SCALE)
и все выставляется правильно

**Reply by nick-nh** [reply]
Есть функция getSLTP_Price

Есть параметр Type. Добавьте в лог вывод данных, чтобы проследить как параметр Type доходит до этой функции. Тогда и найдете причину.

**Comment by 123DNK123** [comment]
Здравствуйте) Возникла очередная проблема. Все настроилось и заработало, но сегодня начало выдавать такую ошибку
attempt to index a nil value (global 'DS')
в  local function getCloseBarTradeState()
    local index = DS:Size() - 1
Помогите пожалуйста. как исправить

**Reply by nick-nh** [reply]
Если Вы вносите изменения в скрипт, то добавляйте логирование в местах получения данных. DS - это поток баров (свечек) по инструменту. Если он не получен, то DS nil. В данном случае надо просто проверить произошло ли корректное получение данных.

**Comment by 123DNK123** [comment]
Здравствуйте. Подскажите пожалуйста. В скрипте реализовано несколько сигналов на покупку и продажу , а также сигналы на выход из сделки дополнительно к Вашим стопам(хорошо себя показали).  Сигналы исполняются, есть необходимость отслеживания исполнения ( какой конкретно сигнал исполнился ). как это можно реализовать , может в текстовом формате. Спасибо

**Reply by nick-nh** [reply]
Не очень понял вопрос.
Все алгоритмы работаю очень просто: есть переменная trend. Когда она меняет значение, то это сигнал. А когда срабатывает стоп, то это просто его контроль. И в логе есть записи о срабатывании стоп заявок и изменении тренда.

**Comment by 123DNK123** [comment]
Тогда это я не разобрался. Извините. Где смотреть лог выполнения скрипта)

**Reply by nick-nh** [reply]
Ну так в самой папке скрипта и сохраняется.

**Comment by ulongMask** [comment]
Спасибо. Много полезностей.

**Comment by 123DNK123** [comment]
Добрый день! сегодня обновилась версия программы QUIK до 10.0.1.18.  После этого скрипт начал выдавать ошибку 
attempt to index a nil value (global 'DS')

local function getCloseBarTradeState()
    local index = DS:Size()-1 

подскажите пожалуйста как исправить

**Reply by nick-nh** [reply]
Добрый день.

Самое простое - проверить, что поток данных заказан, выведя данные в лог.

**Comment by als1313** [comment]
Здравствуйте. 
Спасибо за ваши индикаторы для Quik.
Есть ли возможность убрать из расчета бары выходных дней (суббота и воскресенье) , конкретно интересует индикатор RenkoATR.lua ?

**Comment by nick-nh** [comment]
В теории это можно сделать, но на практике это не такая простая задача, как кажется. Для решения необходимо вводить календарь, т.к. у нас выходной день - это довольно специфическое понятие, их переносят, вводят, убирают, т.е. сказать что это просто суббота и воскресенье нельзя. И даже если пойти самым простым путем и убрать 6, 7 дни, то возникает проблема дырки на графике - индексация ТФ в терминале происходит непрерывно, и если убрать часть баров, то возникает дыра в индексах. Это потребует переработки алгоритма, т.к. сейчас он основан на непрерывной нумерации.

Но в принципе, если индикатор основан на maLib, то можно попробовать модифицировать функцию maLib.CheckIndex, в которой проверяется бар, добавив в нее игнорирование бара с днями 6, 7.

**Reply by als1313** [reply]
Хотел еще уточнить - в индикаторе RenkoATR уровень Revers в некоторых случаях находится на расстоянии равном размеру кирпичика , а в других несколько больше размера кирпичика. Закономерность пока не уловил. Это так задумано или ошибки в работе индикатора ?

**Reply by nick-nh** [reply]
Все же нужен пример, иллюстрация.

**Reply by als1313** [reply]
![RenkoATR](https://github.com/user-attachments/assets/97c4d2f3-a613-413d-b1c4-21036246e7e2)
Вот пример - размер кирпичика вычисляется по ATR

**Comment by nick-nh** [comment]
Это как раз возможно, т.к. при перестроении верхней линии на новый тренд (здесь разворот в лонг) происходит и пересчет размера кирпичика и линий. А линия разворта отстоит на два кирпича от новой верхней линии нового тренда Лонг. Чтобы развернуться необходимо пройти размер двух кирпичиков.

**Reply by als1313** [reply]
Линия реверса должна проходить ниже на два кирпича текущего размера от верхней линии или на один размер кирпича от нижней линии - я правильно понимаю ?

**Reply by nick-nh** [reply]
Если тренд лонг, то два от верхней, один от нижней.

**Reply by als1313** [reply]
Но на приведенной картинке уровень реверса меньше нижней границы почти на два кирпича. Скорее всего расчет уровня реверса идет с учетом "будущего" , еще не сформированного (отрисованного) кирпича , а не по текущим значениям. И при возросшей волатильности , как на картинке , уровень смещается больше.

**Reply by nick-nh** [reply]
Ок. Я проверю, хотя сложно предположить причину данного поведения, т.к. вывод линии реверса банально прост

```
                if trend[index] >= 0 then
                    out_rv = dw[index] - brick[index]
                else
                    out_rv = up[index] + brick[index]
                end
```

Т.е. просто вычитается размер кирпичика от линии. Если линия дальше, значит и размер кирпичика больше.

Приведите полные настройки для данной иллюстрации, запущу проверку с выводом лога на каждый бар.

**Reply by nick-nh** [reply]
Впрочем, причина может быть в методике расчета. При включении режима:
recalc_brick        = 1, -- Пересчитывать размер блока каждый период-бар

размер кирпичика происходит на каждом баре. А линии up и dw пересчитываются только при смене тренда. Так сделано, чтобы линии не скакали при каждой смене. Поэтому если ATR изменился, то и линия реверса сместится.

**Reply by als1313** [reply]
Да  , в настройках recalc_brick = 1 и shift_limit = 1 поэтому линия реверса так строится. А сигналы на продажу и покупку как учитывают линию реверса ?

**Reply by nick-nh** [reply]
Просто когда закрытие его пересекает, то будет сигнал.

**Reply by als1313** [reply]
Возможно ли изменить расчет уровня реверса и соответственно смены тренда с учетом текущего размера кирпича равного разнице up и down ?

**Reply by nick-nh** [reply]
Линия реверса строится от линий кирпичика на расстоянии текущего ATR. В этом плане она строится корректно. Если же Вы хотите, чтобы пересчитывался размер кирпичика при каждом изменении ATR а уже от него и линия реверса, то это подход несколько противоречащий построению Renko. В его классическом варианте, ATR определяется один раз и уже на его базе строится на всем периоде. Что явно не лучший вариант, если необходимо анализировать большой период.

Но и перестраивать размеры кирпича при каждом изменении ATR - это тоже не лучший вариант, т.к. в середине периода кирпичика изменятся его размер. Почему в в середине периода - т.к. пока цена не прошла за реверс, мы внутри него. А период графика, на который нанесен индикатор, будет меньше периода кирпичика.

Впрочем, конечно, это можно сделать, но это уже будет совсем не Renko, а некое подобие NRTR алгоритма.
